Content.Server/Explosions/ExplosionHelper.cs

```diff
@@ -26,129 +32,162 @@ public static class ExplosionHelper
-                if (!entity.Transform.IsMapTransform)
-                    continue;
-
-                if (!entity.Transform.Coordinates.TryDistance(entityManager, coords, out var distance))
+                if (entity.Deleted || !entity.Transform.IsMapTransform)
                 {
                     continue;
                 }
 
-                ExplosionSeverity severity;
-                if (distance < devastationRange)
+                if (!entity.Transform.Coordinates.TryDistance(entityManager, epicenter, out var distance) || distance > maxRange)
                 {
-                    severity = ExplosionSeverity.Destruction;
+                    continue;
                 }
-                else if (distance < heavyImpactRange)
+
+                var severity = CalculateSeverity(distance, devastationRange, heaveyRange);
+                exAct.HandleExplosion(epicenter, entity, severity);
+            }
+        }
+
+        /// <summary>
+        /// Damage tiles inside the range. The type of tile can change depending on a discrete
+        /// damage bracket [light, heavy, devastation], the distance from the epicenter and
+        /// a probabilty bracket [<see cref="LightBreakChance"/>, <see cref="HeavyBreakChance"/>, 1.0].
+        /// </summary>
+        ///
+        private static void DamageTilesInRange(EntityCoordinates epicenter,
+                                               GridId gridId,
+                                               Box2 boundingBox,
+                                               float devastationRange,
+                                               float heaveyRange,
+                                               float maxRange)
+        {
+            var mapManager = IoCManager.Resolve<IMapManager>();
+            if (!mapManager.TryGetGrid(gridId, out var mapGrid))
+            {
+                return;
+            }
+
+            var entityManager = IoCManager.Resolve<IEntityManager>();
+            if (!entityManager.TryGetEntity(mapGrid.GridEntityId, out var grid))
+            {
+                return;
+            }
+
+            _ = grid.TryGetComponent(out GridAtmosphereComponent atmosphereComponent);
+
+            var robustRandom = IoCManager.Resolve<IRobustRandom>();
+            var tileDefinitionManager = IoCManager.Resolve<ITileDefinitionManager>();
+
+            var tilesInGridAndCircle = mapGrid.GetTilesIntersecting(boundingBox);
+
+            foreach (var tile in tilesInGridAndCircle)
+            {
+                var tileLoc = mapGrid.GridTileToLocal(tile.GridIndices);
+                if (!tileLoc.TryDistance(entityManager, epicenter, out var distance) || distance > maxRange)
                 {
-                    severity = ExplosionSeverity.Heavy;
+                    continue;
                 }
-                else if (distance < lightImpactRange)
+
+                if (atmosphereComponent != null && atmosphereComponent.IsAirBlocked(tile.GridIndices))
```

Comment:
Instead of doing this, which is really hacky, you should use `TurfHelpers.IsBlockedTurf` extension on each `TileRef`.